smac_planner/src/a_star.cpp

```diff
@@ -314,6 +331,147 @@ bool AStarAlgorithm<NodeT>::isGoal(NodePtr & node)
   return node == getGoal();
 }
 
+template<>
+AStarAlgorithm<NodeSE2>::NodePtr AStarAlgorithm<NodeSE2>::getAnalyticPath(
+  const NodePtr & node,
+  const NodeGetter & node_getter)
+{
+  ompl::base::ScopedState<> from(node->motion_table.state_space), to(
+    node->motion_table.state_space), s(node->motion_table.state_space);
+  const NodeSE2::Coordinates & node_coords = node->pose;
+  from[0] = node_coords.x;
+  from[1] = node_coords.y;
+  from[2] = node_coords.theta * node->motion_table.bin_size;
+  to[0] = _goal_coordinates.x;
+  to[1] = _goal_coordinates.y;
+  to[2] = _goal_coordinates.theta * node->motion_table.bin_size;
+
+  float d = node->motion_table.state_space->distance(from(), to());
+  NodePtr prev(node);
+  // A move of sqrt(2) is guaranteed to be in a new cell
+  constexpr float sqrt_2 = std::sqrt(2.);
+  unsigned int num_intervals = std::floor(d / sqrt_2);
+
+  // We want to test the nodes on the analytic path out of order.
```

Comment:
This seems maybe a little more complex than required. What we really want is a binary search of the num_points space https://en.wikibooks.org/wiki/Algorithm_Implementation/Search/Binary_search#C++. If we keep halving it and then adding to `visit_order` (rather than the recursive stuff). 

Do you see any improvement in processing time doing it this way?

Reply:
This is effectively a binary search, the difference being that we aren't searching for a particular value - we need to visit every value in the list. So after the first halving, we then need to halve the intervals on both sides and not just one like we would in a regular binary search.

I can't think of a clearer way to do this.

Reply:
Some numbers to consider. These are average over all the failed expansions for a cycle of the planner (around 250 for this test).

"Binary" (ms) | Linear (ms)
-|-
0.008553734513 |	0.007928429204
0.009047308824 | 0.01043161275
0.00695645098 |	0.008020235294
0.005594116564 |	0.005532588957

So it seems like the new approach is generally faster, but not by terribly much and obviously the location of the obstacle plays a part in the results.

Reply:
Is this time for the full expansions? I thought you mentioned that the analytic expansions were pretty heavy, this looks really light weight. 

Given these results, I think its worth the simpler method since its easier to maintain and these numbers are so small that they really don't matter as long as they're not being run on the order of 100+ times

Reply:
In the TB3 world it is typically attempting around 200 expansions before succeeding to find a valid path. I don't think the expansions are as heavy weight any more - I think your optimisations are helping, and the validity and collision checking is very fast now. The maths that OMPL does for finding the analytic paths in the absence of obstacles is fairly straightforward, so it is quick.

I will revert this commit - I agree that it is easier to understand with a linear testing approach and there does not appear to be a clear-cut advantage.